### DISCUSSION TITLE
personal skill not being activated by hotword

### DISCUSSION THREAD

**Discussion by mushie4282** [OpenVoiceOS/ovos-buildroot#240] [Q&A] [answered]
**Goal:** OVOS recognizes "emergency" as a hotword (not an intent) → cellular module texts my phone and verbally says "Contacting..."
**Context:** I used https://github.com/mikejgray/ovos-skill-projen as a baseline my SMS skill. 
**Link to my ORIGINAL skill:** https://github.com/SeniorDesign-RC-LB/skill-r2arc-text
**NEW SKILL THAT WORKS FOR ME** https://github.com/SeniorDesign-RC-LB/ovos-skill-sms

**Comment by JarbasAl** [comment]
you should have some error logs, assuming the [hotword is correctly configured](https://openvoiceos.github.io/ovos-technical-manual/speech_service/#sound-classifiers) etc i see `OVOSSkill.speak_dialog("Notify.911")` but that should be `self.speak_dialog` not `OVOSSkill.speak_dialog`

**Reply by mushie4282** [reply]
I think my configuration is right.
```
(.venv) ovos@raspOvos:~ $ cat ./.config/mycroft/mycroft.conf
{
    "log_level": "INFO",

    "hotwords":{
        "Emergency":{
                "module":"ovos-ww-plugin-vosk",
                "samples":["emergency", "mergency"],
                "rule":"fuzzy",
                "listen":false,
                "active":true,
                "bus_event":"text.emergency"
        }
    }
}
```

**Comment by mushie4282** [comment]


**Reply by builderjer** [reply]
This should have nothing to do with your error, but can be fixed with the latest plugin.
`pip install git+https://github.com/OpenVoiceOS/ovos-PHAL-plugin-mk1`

**Comment by JarbasAl** [comment]
have you confirmed the bus event is sent when the hotword is detected?

if yes, then do you have any errors in the skills service/ovos-core when it is emitted?

**Reply by mushie4282** [reply]
What’s your way of checking if the bus event is sent?

**Comment by mushie4282** [comment]
```
{'type': 'recognizer_loop:hotword', 'data': {'engine': 'VoskWakeWordPlugin', 'sound': None, 'bus_event': 'text.emergency', 'utterance': None, 'stt_lang': 'en-us', 'listen': False, 'wakeup': False, 'stopword': False, 'key_phrase': 'Emergency', 'module': 'ovos-ww-plugin-vosk'}, 'context': {'client_name': 'ovos_dinkum_listener', 'source': 'audio', 'destination': ['skills'], 'lang': 'en-us'}}
```
This is what I've been seeing when I use `mana tail-messagebus` command from `neon-mana-utils`. Is this what you call the bus event being sent out?

**Comment by mushie4282** [comment]
I have been getting these phal_admin messages. 
```
PHAL Admin is started
WARNING - Connection Refused. Is Messagebus Service running?
ERROR - Exception closing websocket at ws://127.0.0.1:8181/core: [Errno 111] Connection refused
WARNING - Message Bus Client will reconnect in 5.0 seconds.
WARNING - Could not send message because connection has closed
ERROR - Exception closing websocket at ws://127.0.0.1:8181/core: Connection to remote host was lost.
WARNING - Message Bus Client will reconnect in 5.0 seconds.
```

**Reply by mushie4282** [reply]
I tried changing the host to 0.0.0.0, which I know is strongly unsecure, but I want to check if it would dismiss the problem. Turns out, it did not.

**Reply by builderjer** [reply]
The above PHAL admin warning/errors happen when the bus is restarted.  You will see one for every time you restarted.

**Reply by mushie4282** [reply]
Shouldn't there be a message that tells the user when the websocket connects? 
`ERROR - Exception closing websocket at ws://127.0.0.1:8181/core: [Errno 111] Connection refused`
Correct me if I'm worng, but the ERROR message strongly suggests that OVOS bus is not working, therefore not bus_events can be emitted.

**Reply by builderjer** [reply]
I have not seen that message. Guess a pr is necessary for that

**Reply by mushie4282** [reply]
Despite the misleading PHAL_admin log messages, I'm able to ping the web socket.
![image](https://github.com/OpenVoiceOS/ovos-buildroot/assets/106356342/a009d84b-cdc3-4833-8b08-f64230457d11)

**Reply by mikejgray** [reply]
ICMP is not the same protocol as TCP, which is what websockets uses. A success from ping doesn't necessary mean the websocket address is serving traffic. I'm wondering if the logs are a red herring - it says it can't disconnect, not connect.

What's the status of the admin PHAL systemd service? sudo systemctl status phal-admin I believe

**Reply by JarbasAl** [reply]
admin phal starst before the bus i think, so it always logs some failures to connect on boot, it eventually reconnects and all works well in principle. but unless you are using a admin plugin you dont even need to care as it's a fully optional component

**Reply by mushie4282** [reply]
`sudo systemctl status phal-admin` isn't a command. Just checked.

**Comment by mushie4282** [comment]
I "reflashed" raspOVOS because there were broken symbolic links to /ramdisk/mycroft/. 
For clarification, in the code snippet below (from OVOS technical manual). When it runs `add_event('recognizer_loop:record_begin',  
                    self.handle_listener_started) `, this means the callback function will be called when the message bus sees "recognizer_loop:record_begin"??
                    
```class ListenForMessageSkill(OVOSSkill):
  def initialize(self):  
      self.add_event('recognizer_loop:record_begin',  
                    self.handle_listener_started)  
      self.add_event('recognizer_loop:record_end',  
                    self.handle_listener_ended)

  def handle_listener_started(self, message):  
      # code to excecute when active listening begins...

  def handle_listener_ended(self, message):  
      # code to excecute when active listening begins...```

**Reply by JarbasAl** [reply]
yup exactly, you can react to any event that way

if something isnt working you should have an error, either on the place that emits the bus event, or in your skill inside the callback after it gets the event

**Reply by mikejgray** [reply]
I don't recall the exact service name. Try systemctl status to get a list, then check the status of the one that says phal

**Comment by mushie4282** [comment]
Thank you guys clarifications! Much appreciated. 
I'm trying to manually launch my skill via `ovos-skill-launcher <skill name or class name>`, but receive the error message below.  I've tried using the class's name and skill name as the arguement.
```
(.venv) ovos@raspOvos:~ $ ovos-skill-launcher TextForHelpSkill
2024-03-08 19:10:11.209 - OVOS - ovos_plugin_manager.templates.audio:<module>:14 - WARNING - Deprecation version=0.1.0. Caller=ovos_plugin_manager.templates:8. ovos_plugin_manager.templates.audio has been deprecated on ovos-audio, move to ovos_plugin_manager.templates.media
2024-03-08 19:10:11.559 - OVOS - ovos_plugin_manager.audio:<module>:7 - WARNING - Deprecation version=0.1.0. Caller=ovos_plugin_manager.plugin_entry:4. ovos_plugin_manager.audio has been deprecated on ovos-audio, move to ovos_plugin_manager.media
2024-03-08 19:10:12.016 - OVOS - mycroft:<module>:37 - WARNING - mycroft has been deprecated! please start importing from ovos_core and companion packages
mycroft module remains available for backwards compatibility and will be removed in version 0.2.0
2024-03-08 19:10:12.112 - OVOS - ovos_workshop.skill_launcher:_connect_to_core:576 - INFO - connected to core
2024-03-08 19:10:12.114 - OVOS - ovos_workshop.skill_launcher:load_skill:592 - INFO - launching skill
Traceback (most recent call last):
  File "/home/ovos/.venv/bin/ovos-skill-launcher", line 8, in <module>
    sys.exit(_launch_script())
             ^^^^^^^^^^^^^^^^
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 654, in _launch_script
    skill.run()
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 602, in run
    self._connect_to_core()
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 577, in _connect_to_core
    self.load_skill()
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 594, in load_skill
    self._launch_plugin_skill()
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 616, in _launch_plugin_skill
    raise ValueError(f"unknown skill_id: {self.skill_id}")
ValueError: unknown skill_id: TextForHelpSkill
(.venv) ovos@raspOvos:~ $
```

**Reply by mikejgray** [reply]
That's not the skill ID. It's going to be constructed in your setup.py script

**Reply by mushie4282** [reply]
I see that skill_id=package_name:SkillClass. I don't see package_name defined in the setup.py.

**Reply by JarbasAl** [reply]
by convention the skill_id is repo.author usually, but can be any string

what you are refering too is just the syntax to tell setup.py what class to load when a skill_id is requested

**Reply by mushie4282** [reply]
After testing out the author's name from the repo's setup.py, I still run into the same unknown skill_id issue.

**Reply by mushie4282** [reply]
I tried `ovos-skill-launcher skill_ovos_news`, but that also resulted in the skill_id issue.

**Reply by mikejgray** [reply]
In an IDE or a Python REPL, instantiate your class and save it to a variable of your choice (`skill`, for example). Then, you can check `skill.skill_id` and see what it's set as. You may also get some error messages that help point out issues.

**Reply by mushie4282** [reply]
I found a quicker method to test if my skill is able to recognize data being sent to the message bus. I copied Sending a Message code snippet from the technical manual and tweaked it to emit the bus event I need to trigger my skill. 
Update: I condensed my skill to only use self.speak(<phrase>). I disconnected the cellular module for testing. 
Update: The code below ACTUALLY trigger my skill!
```
from ovos_bus_client import MessageBusClient, Message

print('Setting up client to connect to a local mycroft instance')
client = MessageBusClient()
client.run_in_thread()

print('Sending speak message...')
client.emit(Message( 'recognizer_loop:hotword', data={'engine': 'VoskWakeWordPlugin', 'sound': None, 'bus_event': 'text.emergency', 'utterance': None, 'stt_lang': 'en-us', 'listen': False, 'wakeup': False, 'stopword': False, 'key_phrase': 'Emergency', 'module': 'ovos-ww-plugin-vosk'}))```

**Comment by mushie4282** [comment]
Since I want my skill to be activated by a defined hotword, would ovos-core still send the message's context to skills? I ask this because based on the image below, I interpreted that ovos-core will emit a message if it recognizes a registered intent, which mine does not have. I pulled the image from https://openvoiceos.github.io/ovos-technical-manual/bus_service/#messages
![image](https://github.com/OpenVoiceOS/ovos-buildroot/assets/106356342/77cd6f33-4d85-4b7e-b0d8-dc6f2ee94f4e)

**Reply by JarbasAl** [reply]
this image is just showing what happens with source/destination as an example, not relevant for you as this is all automatic behind the scenes

what is your issue exactly? is the hotword loading now? is it not emitting the event when detected? is it being loaded but just not detecting your speech as an hotword?

**Reply by mushie4282** [reply]
I'm not able to trigger my skill once dinkum-listener recognizes the hotword. By using `ovos-dinkum-listener`, I saw this message. `2024-03-09 12:10:49.751 - voice - ovos_dinkum_listener.voice_loop.voice_loop:run:233 - INFO - Hotword detected` Therefore, I know the hotword is loaded and being recognized. 
I am able to manually trigger my skill by running a script that imports from ovos_bus_client then emitting a message. If I'm able to manually trigger my skill, then my assuming the problem is the message being emitted by STT. I documented the script in a comment above. (https://github.com/orgs/OpenVoiceOS/discussions/240#discussioncomment-8727444)
For context, my registered event is ` self.add_event('recognizer_loop:hotword', self.handle_text_help)`

**Reply by JarbasAl** [reply]
is this your config? then you should use `self.add_event("text.emergency", ....)`
```
(.venv) ovos@raspOvos:~ $ cat ./.config/mycroft/mycroft.conf
{
    "log_level": "INFO",

    "hotwords":{
        "Emergency":{
                "module":"ovos-ww-plugin-vosk",
                "samples":["emergency", "mergency"],
                "rule":"fuzzy",
                "listen":false,
                "active":true,
                "bus_event":"text.emergency"
        }
    }
}
```

**Reply by mushie4282** [reply]
I updated my "Emergency" config, but generally the same. 
```
"hotwords":{
        "Emergency":{
                "module":"ovos-ww-plugin-vosk",
                "samples":["emergency"],
                 "lang": "en",
                "rule":"fuzzy",
                "listen":false,
                "active":true,
                "bus_event":"text.emergency"
        }
    }
   ```

**Reply by JarbasAl** [reply]
the event in your skill needs to be the one from your config `"bus_event":"text.emergency"`

**Comment by mushie4282** [comment]
Added "text.emergency" to my skill, but it doesn't activate. Could this be an issue related to dinkum-listener emitting a different message than what is configured?
```
def initialize(self):
        self.settings.merge(DEFAULT_SETTINGS, new_only=True)
        self.add_event('recognizer_loop:hotword', self.handle_text_help)
        self.add_event('text.emergency', self.handle_text_help)
        phone = SIM7600X() # instance of cellular module
 ```

**Comment by mushie4282** [comment, marked as answer]
I pivoted my original idea to a more user-friendly way. I'm able to text a phone number via registered intents. Thank you guys for all your help, advice, and guidance! I could have not done it with these discussion posts! 
Here's a small demo video of my new skill. https://youtu.be/Me-3pXEmSA8
